### ISSUE TITLE
Wrong overload resolution with Any union item in argument

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#5249]
This fragment derived from production code causes a false positive:

```py
def f(): pass  # No annotation
t = False
t &= f() and True  # Incompatible types in assignment 
                   # (expression has type "int", variable has type "bool")
```

Here's a self-contained repro:

```py
from typing import overload, Union, Any

class C:
    def f(self, other: C) -> C: ...

class D(C):
    @overload  # type: ignore
    def f(self, other: D) -> D: ...
    @overload
    def f(self, other: C) -> C: ...
    def f(self, other): ...

x: D
y: Union[D, Any]
reveal_type(x.f(y))  # C
```

I'd expect the inferred type for `x.f(y)` to be `Union[D, Any]` (or maybe `D`, though it's slightly less correct arguably).

**Comment by ilevkivskyi** [comment]
I think we should switch to a new union math algorithm, I have some ideas how to make it fast enough to replace the current one (that is apparently to dangerous). I will give it a try now.

**Comment by JukkaL** [comment]
Another real-world example where things go wrong:

```py
from typing import Union, Any, List
i: Union[int, Any]
a: List[int]
reveal_type(a[i])  # Union[int, List[int]]
```

**Comment by JelleZijlstra** [comment]
I think that one is actually correct: if `i` is a slice object, it will return `List[int]`.

**Comment by JukkaL** [comment]
`List[int]` is inconsistent with how `Any` works with overloads generally -- if there is ambiguity due to `Any` types in overload resolution, we tend to fall back to `Any` to avoid false positives. False negatives are preferred over false positives when `Any` types are involved. Here is a simple analogy:

```py
from typing import Union, Any
class A:
    x: str
class B:
    x: int
a: Union[A, Any]
reveal_type(a.x)  # Union[str, Any]
```

We don't infer `Union[str, int]` as the type of `a.x` even though `int` is a plausible type for `a.x` -- since it could well be something else. Similarly, the precise type of `i` could actually be just `int` (though we can't tell for certain), and assuming that it can be a slice would be wrong.

**Comment by ilevkivskyi** [comment]
The PR I am working will produce (after some tweaks that I didn't push yet) `Union[int, Any]` for this example:
```python
from typing import Union, Any, List
i: Union[int, Any]
a: List[int]
reveal_type(a[i]) # Union[int, Any]
```
@JukkaL does this make sense to you?

**Comment by ilevkivskyi** [comment]
(Also I think `Union[D, Any]` in the original example is best.)

**Comment by JukkaL** [comment]
Yes, I think that`Union[int, Any]` is the most correct type for `a[i]`, and similarly `Union[D, Any]` for the other example.
